File: packages/web/src/lib/pricing.ts
import type { UsageTokens } from "@/lib/api";

export type ModelRate = {
  input: number; // USD per 1M tokens
  output: number; // USD per 1M tokens
  reasoning: number; // USD per 1M tokens
  cacheRead: number; // USD per 1M tokens
  cacheWrite: number; // USD per 1M tokens
  label?: string;
};

/**
 * Model API pricing table per million tokens (USD).
 */
export function getModelRate(provider: string, model: string): ModelRate {
  const p = (provider || "").toLowerCase();
  const m = (model || "").toLowerCase();

  // Normalize model name (remove provider prefixes if present, e.g. "anthropic/claude-3-7-sonnet")
  const cleanModel = m.includes("/") ? m.split("/").pop() || m : m;

  // 1. Anthropic / Claude
  if (p === "claude" || p.includes("anthropic") || cleanModel.includes("claude")) {
    if (cleanModel.includes("opus")) {
      return { input: 15.0, output: 75.0, reasoning: 75.0, cacheRead: 1.5, cacheWrite: 18.75, label: "Claude Opus ($15/$75 per 1M)" };
    }
    if (cleanModel.includes("haiku")) {
      return { input: 0.8, output: 4.0, reasoning: 4.0, cacheRead: 0.08, cacheWrite: 1.0, label: "Claude Haiku ($0.80/$4.00 per 1M)" };
    }
    // Sonnet (3.5, 3.7, 4, 5)
    return { input: 3.0, output: 15.0, reasoning: 15.0, cacheRead: 0.3, cacheWrite: 3.75, label: "Claude Sonnet ($3/$15 per 1M)" };
  }

  // 2. OpenAI / Codex
  if (p === "codex" || p.includes("openai") || cleanModel.includes("gpt") || cleanModel.startsWith("o1") || cleanModel.startsWith("o3") || cleanModel.startsWith("o4")) {
    if (cleanModel.includes("o1-mini") || cleanModel.includes("o3-mini") || cleanModel.includes("o4-mini")) {
      return { input: 1.1, output: 4.4, reasoning: 4.4, cacheRead: 0.55, cacheWrite: 1.1, label: "o3-mini ($1.10/$4.40 per 1M)" };
    }
    if (cleanModel.startsWith("o1") || cleanModel.startsWith("o3") || cleanModel.startsWith("o4")) {
      return { input: 15.0, output: 60.0, reasoning: 60.0, cacheRead: 7.5, cacheWrite: 15.0, label: "o1 / o3 ($15/$60 per 1M)" };
    }
    if (cleanModel.includes("4o-mini")) {
      return { input: 0.15, output: 0.6, reasoning: 0.6, cacheRead: 0.075, cacheWrite: 0.15, label: "GPT-4o-mini ($0.15/$0.60 per 1M)" };
    }
    if (cleanModel.includes("4o")) {
      return { input: 2.5, output: 10.0, reasoning: 10.0, cacheRead: 1.25, cacheWrite: 2.5, label: "GPT-4o ($2.50/$10 per 1M)" };
    }
    if (cleanModel.includes("gpt-5")) {
      return { input: 3.0, output: 15.0, reasoning: 15.0, cacheRead: 1.5, cacheWrite: 3.0, label: "GPT-5 ($3/$15 per 1M)" };
    }
    if (cleanModel.includes("gpt-4")) {
      return { input: 10.0, output: 30.0, reasoning: 30.0, cacheRead: 5.0, cacheWrite: 10.0, label: "GPT-4 Turbo ($10/$30 per 1M)" };
    }
    if (cleanModel.includes("3.5")) {
      return { input: 0.5, output: 1.5, reasoning: 1.5, cacheRead: 0.25, cacheWrite: 0.5, label: "GPT-3.5 Turbo ($0.50/$1.50 per 1M)" };
    }
    return { input: 2.5, output: 10.0, reasoning: 10.0, cacheRead: 1.25, cacheWrite: 2.5, label: "GPT-4o ($2.50/$10 per 1M)" };
  }

  // 3. Google Gemini / Antigravity
  if (p === "agy" || p.includes("gemini") || p.includes("google") || cleanModel.includes("gemini")) {
    if (cleanModel.includes("flash-lite")) {
      return { input: 0.0375, output: 0.15, reasoning: 0.15, cacheRead: 0.01, cacheWrite: 0.0375, label: "Gemini Flash Lite ($0.0375/$0.15 per 1M)" };
    }
    if (cleanModel.includes("flash")) {
      return { input: 0.075, output: 0.3, reasoning: 0.3, cacheRead: 0.01875, cacheWrite: 0.075, label: "Gemini Flash ($0.075/$0.30 per 1M)" };
    }
    if (cleanModel.includes("pro") || cleanModel.includes("ultra")) {
      return { input: 1.25, output: 5.0, reasoning: 5.0, cacheRead: 0.3125, cacheWrite: 1.25, label: "Gemini Pro ($1.25/$5.00 per 1M)" };
    }
    return { input: 0.5, output: 2.0, reasoning: 2.0, cacheRead: 0.125, cacheWrite: 0.5, label: "Gemini ($0.50/$2.00 per 1M)" };
  }

  // 4. DeepSeek
  if (p.includes("deepseek") || cleanModel.includes("deepseek")) {
    if (cleanModel.includes("reasoner") || cleanModel.includes("r1")) {
      return { input: 0.55, output: 2.19, reasoning: 2.19, cacheRead: 0.14, cacheWrite: 0.55, label: "DeepSeek R1 ($0.55/$2.19 per 1M)" };
    }
    return { input: 0.14, output: 0.28, reasoning: 0.28, cacheRead: 0.014, cacheWrite: 0.14, label: "DeepSeek V3 ($0.14/$0.28 per 1M)" };
  }

  // 5. xAI / Grok
  if (p === "grok" || p.includes("x-ai") || cleanModel.includes("grok")) {
    if (cleanModel.includes("mini")) {
      return { input: 0.2, output: 1.0, reasoning: 1.0, cacheRead: 0.05, cacheWrite: 0.2, label: "Grok Mini ($0.20/$1.00 per 1M)" };
    }
    return { input: 2.0, output: 10.0, reasoning: 10.0, cacheRead: 0.5, cacheWrite: 2.0, label: "Grok ($2/$10 per 1M)" };
  }

  // 6. Copilot
  if (p === "copilot") {
    if (cleanModel.includes("claude")) {
      return { input: 3.0, output: 15.0, reasoning: 15.0, cacheRead: 0.3, cacheWrite: 3.75, label: "Copilot / Claude ($3/$15 per 1M)" };
    }
    if (cleanModel.includes("mini")) {
      return { input: 0.15, output: 0.6, reasoning: 0.6, cacheRead: 0.075, cacheWrite: 0.15, label: "Copilot / 4o-mini ($0.15/$0.60 per 1M)" };
    }
    return { input: 2.5, output: 10.0, reasoning: 10.0, cacheRead: 1.25, cacheWrite: 2.5, label: "Copilot / GPT-4o ($2.50/$10 per 1M)" };
  }

  // Balanced fallback (Mid-tier $2.00 in / $8.00 out per 1M)
  return { input: 2.0, output: 8.0, reasoning: 8.0, cacheRead: 0.4, cacheWrite: 2.0, label: "Standard Rate ($2/$8 per 1M)" };
}

/**
 * Calculates estimated cost for given tokens and model rate.
 */
export function estimateTokenCost(provider: string, model: string, tokens: UsageTokens): number {
  const rate = getModelRate(provider, model);
  const inputCost = (tokens.input / 1_000_000) * rate.input;
  const outputCost = (tokens.output / 1_000_000) * rate.output;
  const reasoningCost = (tokens.reasoning / 1_000_000) * (rate.reasoning || rate.output);
  const cacheReadCost = (tokens.cache_read / 1_000_000) * rate.cacheRead;
  const cacheWriteCost = (tokens.cache_write / 1_000_000) * rate.cacheWrite;

  return inputCost + outputCost + reasoningCost + cacheReadCost + cacheWriteCost;
}

File: packages/cli/internal/agentusage/pricing.go
package agentusage

import (
	"strings"
)

// ModelRate defines the pricing per 1 Million tokens (in USD).
type ModelRate struct {
	Input      float64
	Output     float64
	Reasoning  float64
	CacheRead  float64
	CacheWrite float64
}

// EstimateCost calculates the estimated USD cost for a given provider, model,
// and token distribution based on official provider API rates.
func EstimateCost(provider, model string, t Tokens) float64 {
	rate := lookupRate(provider, model)

	// Price = (Tokens / 1,000,000) * RatePerMillion
	inputCost := (float64(t.Input) / 1_000_000.0) * rate.Input
	outputCost := (float64(t.Output) / 1_000_000.0) * rate.Output

	// Reasoning tokens bill at output rates unless specified differently
	reasoningRate := rate.Reasoning
	if reasoningRate == 0 {
		reasoningRate = rate.Output
	}
	reasoningCost := (float64(t.Reasoning) / 1_000_000.0) * reasoningRate

	// Cache read is typically 10% - 50% of input rate
	cacheReadRate := rate.CacheRead
	if cacheReadRate == 0 && rate.Input > 0 {
		cacheReadRate = rate.Input * 0.10
	}
	cacheReadCost := (float64(t.CacheRead) / 1_000_000.0) * cacheReadRate

	// Cache write is typically 1.25x of input rate
	cacheWriteRate := rate.CacheWrite
	if cacheWriteRate == 0 && rate.Input > 0 {
		cacheWriteRate = rate.Input * 1.25
	}
	cacheWriteCost := (float64(t.CacheWrite) / 1_000_000.0) * cacheWriteRate

	return inputCost + outputCost + reasoningCost + cacheReadCost + cacheWriteCost
}

func lookupRate(provider, model string) ModelRate {
	p := strings.ToLower(provider)
	m := strings.ToLower(model)

	// Normalize model name (strip provider prefixes if embedded like "anthropic/claude-3-7-sonnet")
	if idx := strings.LastIndex(m, "/"); idx != -1 {
		m = m[idx+1:]
	}

	// ── 1. Anthropic / Claude ────────────────────────────────────────────────
	if p == "claude" || strings.Contains(p, "anthropic") || strings.Contains(m, "claude") {
		switch {
		case strings.Contains(m, "opus"):
			return ModelRate{Input: 15.00, Output: 75.00, Reasoning: 75.00, CacheRead: 1.50, CacheWrite: 18.75}
		case strings.Contains(m, "haiku"):
			return ModelRate{Input: 0.80, Output: 4.00, Reasoning: 4.00, CacheRead: 0.08, CacheWrite: 1.00}
		case strings.Contains(m, "sonnet"):
			// Sonnet 3.5 / 3.7 / 4 / 5
			return ModelRate{Input: 3.00, Output: 15.00, Reasoning: 15.00, CacheRead: 0.30, CacheWrite: 3.75}
		default:
			return ModelRate{Input: 3.00, Output: 15.00, Reasoning: 15.00, CacheRead: 0.30, CacheWrite: 3.75}
		}
	}

	// ── 2. OpenAI / Codex ────────────────────────────────────────────────────
	if p == "codex" || strings.Contains(p, "openai") || strings.Contains(m, "gpt") || strings.Contains(m, "o1") || strings.Contains(m, "o3") {
		switch {
		case strings.Contains(m, "o1-mini") || strings.Contains(m, "o3-mini") || strings.Contains(m, "o4-mini"):
			return ModelRate{Input: 1.10, Output: 4.40, Reasoning: 4.40, CacheRead: 0.55, CacheWrite: 1.10}
		case strings.Contains(m, "o1") || strings.Contains(m, "o3"):
			return ModelRate{Input: 15.00, Output: 60.00, Reasoning: 60.00, CacheRead: 7.50, CacheWrite: 15.00}
		case strings.Contains(m, "4o-mini"):
			return ModelRate{Input: 0.15, Output: 0.60, Reasoning: 0.60, CacheRead: 0.075, CacheWrite: 0.15}
		case strings.Contains(m, "4o"):
			return ModelRate{Input: 2.50, Output: 10.00, Reasoning: 10.00, CacheRead: 1.25, CacheWrite: 2.50}
		case strings.Contains(m, "gpt-5"):
			return ModelRate{Input: 3.00, Output: 15.00, Reasoning: 15.00, CacheRead: 1.50, CacheWrite: 3.00}
		case strings.Contains(m, "gpt-4-turbo") || strings.Contains(m, "gpt-4"):
			return ModelRate{Input: 10.00, Output: 30.00, Reasoning: 30.00, CacheRead: 5.00, CacheWrite: 10.00}
		case strings.Contains(m, "3.5"):
			return ModelRate{Input: 0.50, Output: 1.50, Reasoning: 1.50, CacheRead: 0.25, CacheWrite: 0.50}
		default:
			return ModelRate{Input: 2.50, Output: 10.00, Reasoning: 10.00, CacheRead: 1.25, CacheWrite: 2.50}
		}
	}

	// ── 3. Google Gemini / Antigravity (agy) ──────────────────────────────────
	if p == "agy" || strings.Contains(p, "gemini") || strings.Contains(p, "google") || strings.Contains(m, "gemini") {
		switch {
		case strings.Contains(m, "flash-lite"):
			return ModelRate{Input: 0.0375, Output: 0.15, Reasoning: 0.15, CacheRead: 0.01, CacheWrite: 0.0375}
		case strings.Contains(m, "flash"):
			return ModelRate{Input: 0.075, Output: 0.30, Reasoning: 0.30, CacheRead: 0.01875, CacheWrite: 0.075}
		case strings.Contains(m, "pro") || strings.Contains(m, "ultra"):
			return ModelRate{Input: 1.25, Output: 5.00, Reasoning: 5.00, CacheRead: 0.3125, CacheWrite: 1.25}
		default:
			return ModelRate{Input: 0.50, Output: 2.00, Reasoning: 2.00, CacheRead: 0.125, CacheWrite: 0.50}
		}
	}

	// ── 4. DeepSeek ──────────────────────────────────────────────────────────
	if strings.Contains(p, "deepseek") || strings.Contains(m, "deepseek") {
		switch {
		case strings.Contains(m, "reasoner") || strings.Contains(m, "r1"):
			return ModelRate{Input: 0.55, Output: 2.19, Reasoning: 2.19, CacheRead: 0.14, CacheWrite: 0.55}
		default:
			return ModelRate{Input: 0.14, Output: 0.28, Reasoning: 0.28, CacheRead: 0.014, CacheWrite: 0.14}
		}
	}

	// ── 5. xAI / Grok ────────────────────────────────────────────────────────
	if p == "grok" || strings.Contains(p, "x-ai") || strings.Contains(m, "grok") {
		switch {
		case strings.Contains(m, "mini"):
			return ModelRate{Input: 0.20, Output: 1.00, Reasoning: 1.00, CacheRead: 0.05, CacheWrite: 0.20}
		default:
			return ModelRate{Input: 2.00, Output: 10.00, Reasoning: 10.00, CacheRead: 0.50, CacheWrite: 2.00}
		}
	}

	// ── 6. Copilot ───────────────────────────────────────────────────────────
	if p == "copilot" {
		switch {
		case strings.Contains(m, "claude"):
			return ModelRate{Input: 3.00, Output: 15.00, Reasoning: 15.00, CacheRead: 0.30, CacheWrite: 3.75}
		case strings.Contains(m, "gpt-4o-mini"):
			return ModelRate{Input: 0.15, Output: 0.60, Reasoning: 0.60, CacheRead: 0.075, CacheWrite: 0.15}
		default:
			return ModelRate{Input: 2.50, Output: 10.00, Reasoning: 10.00, CacheRead: 1.25, CacheWrite: 2.50}
		}
	}

	// Default fallback (balanced mid-tier rates: $2/1M in, $8/1M out)
	return ModelRate{Input: 2.00, Output: 8.00, Reasoning: 8.00, CacheRead: 0.40, CacheWrite: 2.00}
}
